accepted/2020/better-obsoletion/securestring-obsoletion.md

```diff
@@ -0,0 +1,221 @@
+
+```cs
+// Example showing unshrouding a SensitiveData<char> into a temp char[].
+SensitiveData<char> sensitiveData = GetSensitiveStringData();
+char[] tempArray = GC.AllocateArray<char>(sensitiveData.Length, pinned: true);
+try
+{
+    sensitiveData.CopyTo(tempArray);
+    // use 'tempArray' here
+}
+finally
+{
+    Array.Clear(tempArray);
+}
+
+// Example showing unshrouding a SensitiveData<char> into a string.
+SensitiveData<char> sensitiveData = GetSensitiveStringData();
+string plaintextString = string.Create(
+    sensitiveData.Length, sensitiveData, (span, sensitiveData) =>
+    {
+        sensitiveData.CopyTo(span);
+    });
+```
+
+We would clearly document that this type is not intended to offer cryptographic protection of the buffer contents; it's simply intended to enforce more hygienic usage of data by callers.
+
+The type also intentionally does not give callers access to the raw backing span. This is because how the backing data is stored is strictly an implementation detail (it might not even be in-proc!), and exposing the span would leak that detail into the public API surface. The `CopyTo` method is the only way to get the raw contents. This also means that such buffers are immutable once created.
+
+__We should take care not to position this "sensitive buffer" type as a security feature for the purpose of achieving compliance.__ It's intended solely for code hygiene purposes and to avoid accidental disclosure. Scenarios like achieving PCI compliance require more fundamental runtime work, such as that under consideration at https://github.com/dotnet/runtime/issues/10480.
+
+This proposal intentionally excludes `ToString`, `ToArray`, or other methods which might unshroud the buffer. `ToString` is called frequently by library code as part of logging, and we always want unshrouding to be a _deliberate_ action. Additionally, we want the caller to maintain full control over the unshrouded buffer where the contents are written, which precludes "opinionated" APIs like `ToArray` which create such buffers themselves.
+
+### Best practices for implementation
+
+If the backing buffer for the sensitive data is stored in-proc, we should attempt to use a separate unmanaged heap for storing these values, and we should not hand out direct references to this heap. This helps mitigate the possibility of sensitive data being leaked due to corruption of the normal heap. This concept is [already in use by OpenSSL](https://www.openssl.org/docs/man1.1.0/man3/CRYPTO_secure_malloc_initialized.html) and a few high-sensitivity Windows services. If using a separate heap is not feasible, we should attempt to use .NET's pinned object heap.
```

Comment:
> If the backing buffer for the sensitive data is stored in-proc, we should attempt to use a separate unmanaged heap for storing these values, and we should not hand out direct references to this heap.

It's not clear to me how we'd deal with how this data gets into the process in the first place, such as:

* Microsoft.Extensions.Configuration
* Environment variables

For example, we should consider a method like `SensitiveData<char> Environment.GetSensitveEnvironmentVariable(string name)`.

Reply:
What's wrong with the following? It's short and simple and doesn't require API changes beyond what's proposed here.

```cs
string envVar = Enviroment.GetEnvironmentVariable(name);
SensitiveData<char> sensitiveData = new SensitiveData<char>(envVar);
```

Remember, the point of this type is not to prevent the initial materialization of strings / arrays, nor is it to hide the contents of this buffer from an active memory dump. The point is to mark a piece of data as "sensitive" and to prevent _accidental_ re-materialization of this string / array. The code sample above fits with this spirit.

Reply:
> Remember, the point of this type is not to prevent the initial materialization of strings / arrays, nor is it to hide the contents of this buffer from an active memory dump. 

Which is good, since the entire environment block is already in dumpable memory :smile:

Reply:
> Remember, the point of this type is not to prevent the initial materialization of strings / arrays, nor is it to hide the contents of this buffer from an active memory dump. The point is to mark a piece of data as "sensitive" and to prevent _accidental_ re-materialization of this string / array. The code sample above fits with this spirit.

If we believe that to be the case, then I this certainly simplifies the design.

Reply:
> Which is good, since the entire environment block is already in dumpable memory

there is prior art for redacting dumps, eg., the triage dumps that blank out much of the data leaving only the stack. in theory, if they don't already, they could clean the environment block. but they couldn't clean an arbitrary part of the GC heap. so I'm not sure the analogy with the env block holds entirely?